theano/tensor/basic.py

```diff
@@ -5902,24 +5902,74 @@ class Diagonal(Op):
 
     """
     __props__ = ("offset", "axis1", "axis2")
+    default_offset = 0
+    default_axis1 = 0
+    default_axis2 = 1
 
     def __init__(self, offset=0, axis1=0, axis2=1):
-        if numpy_diagonal_return_view:
+        self.view = view
+        if self.view and not numpy_diagonal_return_view:
+            warnings.warn("View will forced to False. Diagonal property view is "
+                          "set to True but numpy version %s and prior versions of "
+                          "numpy.diagonal() do not return a view. Update "
+                          "numpy to use Diagonal(view=True)" % 
+                          numpy.version.version)
+            self.view = False
+        if self.view:
             self.view_map = {0: [0]}
         self.offset = offset
         self.axis1 = axis1
         self.axis2 = axis2
 
-    def make_node(self, x):
-        x = as_tensor_variable(x)
-        assert x.ndim >= 2
-        return Apply(self, [x], [tensor(dtype=x.dtype,
-                                        broadcastable=[False] * (x.ndim - 1))])
+    def make_node(self, _x):
+        if not isinstance(_x, theano.Variable):
+            x = as_tensor_variable(_x)
+        else:
+            x = _x
+
+        if x.ndim < 2:
+            raise ValueError('Diagonal needs an input with 2 or more '
+                             'dimensions', x)
+        return Apply(self, [x], [x.type.__class__(
+            dtype=x.dtype,
+            broadcastable=[False] * (x.ndim - 1))()])
 
     def perform(self, node, inputs, outputs):
         (x,) = inputs
         (z,) = outputs
-        z[0] = x.diagonal(self.offset, self.axis1, self.axis2)
+        # zero-dimensional matrices ...
```

Comment:
Ah yes, we should test if it returns numpy view, and do diagonal
accordingly.

On Tuesday, 26 April 2016, Xavier Bouthillier <notifications@github.com>
wrote:

> In theano/tensor/basic.py
> <https://github.com/Theano/Theano/pull/4424#discussion_r61189011>:
>
> >
> >      def perform(self, node, inputs, outputs):
> >          (x,) = inputs
> >          (z,) = outputs
> > -        z[0] = x.diagonal(self.offset, self.axis1, self.axis2)
> > +        # zero-dimensional matrices ...
>
> Until 1.7 and then is was a view, right?
>
>
> http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.diagonal.html
>
> We will keep the code that verifies if numpy's diagonal is a copy or a
> view, and force self.view=False in the first case.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/4424/files/c1a864ccf486d74df2625c15a67ded78c5b85cdf#r61189011>
>


-- 
LIN, Zhouhan
Department of Computer Science and Operational Research
University of Montreal

Phone: +01-514-586-0551
Email: lin.zhouhan@gmail.com
Homepage:http://zhouhanlin.webs.com/